### PR TITLE
[#32] Add extension points for new private data collection types #32

### PR DESCRIPTION
**Pull request by fqutishat**
Signed-off-by: Bob Stasyszyn <Bob.Stasyszyn@securekey.com>
Signed-off-by: Firas Qutishat <firas.qutishat@securekey.com>

#32

### CODE REVIEW

File: core/peer/peer_test.go [hunk 1]
```diff
@@ -23,6 +23,7 @@ import (
 	validation "github.com/hyperledger/fabric/core/handlers/validation/api"
 	"github.com/hyperledger/fabric/core/ledger/mock"
 	ledgermocks "github.com/hyperledger/fabric/core/ledger/mock"
+	kmocks "github.com/hyperledger/fabric/extensions/mocks"
```
Comment (troyronda) [inline core/peer/peer_test.go]: kmocks?

**Comment by troyronda** [comment]
There are potentially a few changes in this PR that could have more test coverage.

File: core/peer/peer_test.go (hunk 1 above)
Comment (fqutishat) [reply core/peer/peer_test.go]: I remove it

File: core/ledger/kvledger/txmgmt/txmgr/lockbasedtxmgr/collection_val_ext_test.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+	"testing"
+
+	"github.com/hyperledger/fabric/core/ledger"
+	"github.com/hyperledger/fabric/core/ledger/kvledger/txmgmt/version"
+	"github.com/hyperledger/fabric/core/ledger/mock"
+	"github.com/hyperledger/fabric/protos/common"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestTransientCollectionValidation(t *testing.T) {
```
Comment (bstasyszyn) [inline core/ledger/kvledger/txmgmt/txmgr/lockbasedtxmgr/collection_val_ext_test.go]: I think we should remove the word 'transient'. This test is to make sure that no error is returned for unknown collection types.

File: core/ledger/kvledger/txmgmt/txmgr/lockbasedtxmgr/collection_val_ext_test.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	const key2 = "key1"
+
+	_, err = sim.GetPrivateData("ns1", "coll1", key1)
+	assert.NoError(t, err)
+
+	_, err = sim.GetPrivateDataMultipleKeys("ns1", "coll1", []string{key1, key2})
+	assert.NoError(t, err)
+}
+
+func populateTransientCollConfigForTest(t *testing.T, txMgr *LockBasedTxMgr, nsColls []collConfigkey, ht *version.Height) {
```
Comment (bstasyszyn) [inline core/ledger/kvledger/txmgmt/txmgr/lockbasedtxmgr/collection_val_ext_test.go]: Rename to populateUnknownCollConfigForTest.

File: core/ledger/kvledger/txmgmt/txmgr/lockbasedtxmgr/collection_val_ext_test.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+		ns, coll := nsColl.ns, nsColl.coll
+		pkg, ok := m[ns]
+		if !ok {
+			pkg = &common.CollectionConfigPackage{}
+			m[ns] = pkg
+		}
+		tCollConfig := &common.CollectionConfig_StaticCollectionConfig{
+			StaticCollectionConfig: &common.StaticCollectionConfig{
+				Name: coll,
+				Type: common.CollectionType_COL_TRANSIENT,
```
Comment (bstasyszyn) [inline core/ledger/kvledger/txmgmt/txmgr/lockbasedtxmgr/collection_val_ext_test.go]: Use CollectionType_COL_UNKNOWN.

File: peer/chaincode/common.go [hunk 5]
```diff
@@ -180,6 +180,14 @@ func chaincodeInvokeOrQuery(cmd *cobra.Command, invoke bool, cf *ChaincodeCmdFac
 	return nil
 }
 
+// CollectionType enumerates the various types of private data collections.
```
Comment (bstasyszyn) [inline peer/chaincode/common.go]: I don't this this change should be part of this commit since it references transient data.

File: peer/chaincode/transientcollconfig_test.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
+/*
```
Comment (bstasyszyn) [inline peer/chaincode/transientcollconfig_test.go]: These changes shouldn't be part of this commit.

File: peer/node/start.go [hunk 7]
```diff
@@ -193,6 +195,14 @@ func serve(args []string) error {
 
 	lifecycleCache := lifecycle.NewCache(lifecycleResources, mspID)
 
+	transientDataProvider := collretriever.NewProvider(
```
Comment (bstasyszyn) [inline peer/node/start.go]: Rename var to collDataProvider

File: protos/common/collection.proto [hunk 8]
```diff
@@ -13,6 +13,13 @@ option java_package = "org.hyperledger.fabric.protos.common";
 
 package common;
 
+// CollectionType enumerates the various types of private data collections.
+enum CollectionType {
+    COL_UNKNOWN = 0;    // Unspecified type - will use PRIVATE as default.
+    COL_PRIVATE = 1;    // Persisted private data collection.
+    COL_TRANSIENT = 2;  // Transient private data collection.
```
Comment (bstasyszyn) [inline protos/common/collection.proto]: Remove TRANSIENT_DATA

File: protos/common/collection.proto [hunk 9]
```diff
@@ -61,6 +68,14 @@ message StaticCollectionConfig {
...
     // access logic in the chaincode)
     bool member_only_write = 7;
+    // The type of collection.
+    CollectionType type = 9900;
+    // The time after which the collection data expires. For example,
+    // if the value is set to "10m" then the data will be purged
+    // 10 minutes after it was stored. An empty value indicates that
+    // the data should never be purged.
+    // The format of this string must be parseable by time.ParseDuration
+    string time_to_live = 9901;
```
Comment (bstasyszyn) [inline protos/common/collection.proto]: Remove time_to_live

File: core/ledger/kvledger/txmgmt/txmgr/lockbasedtxmgr/collection_val_ext_test.go (hunk 2 above)
Comment (fqutishat) [reply core/ledger/kvledger/txmgmt/txmgr/lockbasedtxmgr/collection_val_ext_test.go]: done

File: core/ledger/kvledger/txmgmt/txmgr/lockbasedtxmgr/collection_val_ext_test.go (hunk 3 above)
Comment (fqutishat) [reply core/ledger/kvledger/txmgmt/txmgr/lockbasedtxmgr/collection_val_ext_test.go]: done

File: core/ledger/kvledger/txmgmt/txmgr/lockbasedtxmgr/collection_val_ext_test.go (hunk 4 above)
Comment (fqutishat) [reply core/ledger/kvledger/txmgmt/txmgr/lockbasedtxmgr/collection_val_ext_test.go]: done

File: peer/node/start.go (hunk 7 above)
Comment (fqutishat) [reply peer/node/start.go]: done

**Comment by fqutishat** [comment]
verify

**Comment by fqutishat** [comment]
verify
